Enable extensions/v1beta1 by default and allow runtimeConfig to support enabling/disabling specific resources

nikhiljindal opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/15403

List of changes:
- Enable extensions/v1beta1 by default. Only jobs, ingress and Horizontal Pod Autoscaler are enabled by default. All other extensions resources are disabled.
- Added support to runtime-config to specify overrides to enable/disable individual resources in addition to whole api group versions.

Example of valid runtime-config values:

`runtime-config="extensions/v1beta1"` enables all extension resources.
`runtime-config="extensions/v1beta1,extensions/v1beta1/jobs=false"` adds an override to disable  jobs.
`runtime-config="extensions/v1beta1,extension/v1beta1/deployments=true"` adds an override to enable deployment.
`runtime-config="extensions/v1beta1,extension/v1beta1/jobs=false,extension/v1beta1/deployments=true"` adds 2 overrides (enable deployments and disable jobs)

The resource specific enabling/disabling is only supported for extensions.

nikhiljindal committed:
Allowing runtimeConfig to support enabling/disabling specific extension resources

nikhiljindal commented:
cc @lavalamp @caesarxuchao

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -658,3 +669,19 @@ func (s *APIServer) getRuntimeConfigValue(apiKey string, defaultValue bool) bool
 	return defaultValue
 }
+
+func (s *APIServer) getEnabledAndDisabledExtensionResources(extensionsPrefix string) (map[string]struct{}, map[string]struct{}) {
```
`getResourceOverrides(groupVersion string) map[string]bool {`

nikhiljindal replied:
Removed

caesarxuchao commented on cmd/kube-apiserver/app/server_test.go:
```diff
@@ -154,3 +154,55 @@ func TestUpdateEtcdOverrides(t *testing.T) {
+				prefix + "foo": "",
+			},
+			enabledResources: map[string]struct{}{
+				"foo": struct{}{},
```
this looks counter-intuitive to me. I think default to "true" is better.

nikhiljindal replied:
removed

lavalamp commented on pkg/master/master.go:
```diff
@@ -182,8 +182,10 @@ type Config struct {
-	EnableExp bool
+	DisableV1                 bool
+	EnableExtensions          bool
+	EnableExtensionResources  map[string]struct{} // If present, only these resources are enabled in the extensions api group, all others are disabled. This is ignored if EnableExtensions is false.
```
ExtensionOverrides map[string]bool

nikhiljindal replied:
Are you suggesting combining Enable and Disable maps or just changing each of them to map[string]bool?

lavalamp replied:
Sorry for not being clear-- I meant combining them. I'm pretty sure it's
got some odd behavior right now if you both specifically enable and disable
some resources.

On Mon, Oct 12, 2015 at 6:15 PM, Nikhil Jindal <notifications@github.com>
wrote:

> In pkg/master/master.go
> <https://github.com/kubernetes/kubernetes/pull/15516#discussion_r41818286>
> :
>
> > @@ -182,8 +182,10 @@ type Config struct {
> >  	// allow downstream consumers to disable swagger
> >  	EnableSwaggerSupport bool
> >  	// allow api versions to be conditionally disabled
> > -	DisableV1 bool
> > -	EnableExp bool
> > +	DisableV1                 bool
> > +	EnableExtensions          bool
> > +	EnableExtensionResources  map[string]struct{} // If present, only these resources are enabled in the extensions api group, all others are disabled. This is ignored if EnableExtensions is false.
>
> Are you suggesting combining Enable and Disable maps or just changing each
> of them to map[string]bool?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15516/files#r41818286>.
>

nikhiljindal replied:
Note that only one of enableExtensionResources or disableExtensionResources can be non-empty.
By keeping them separate, checking that is easy (len(enableExtensionResources) == 0 || len(disableExtensionResources) == 0). Else, we will have to iterate through all elements and then see that all of them have the same bool value.

lavalamp replied:
Maybe we can return a default value and a list/set of overrides, then?

lavalamp replied:
(and if it has to be a set, please use util/sets.String)

nikhiljindal replied:
Yes thanks for pointing that out, I hadnt seen it. Will switch to util.Sets.

nikhiljindal replied:
compressed all these variables into a single apiGroupVersionOverrides

caesarxuchao commented:
SGTM. Thank you @nikhiljindal!

nikhiljindal commented:
Rebased

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -556,7 +565,9 @@ func (s *APIServer) Run(_ []string) error {
 		AdmissionControl:          admissionController,
 		DisableV1:                 disableV1,
-		EnableExp:                 enableExp,
+		EnableExtensions:          enableExtensions,
```
It would be really nice if this could be a map/list of groups with a list of versions with a list of resources. We can't keep making variables here for every group/version.

nikhiljindal replied:
Yes I can make this a generic struct:
```
ResourceOverrides struct {
  groupVersion string,
  resources util.Set,
  isEnableList bool
}
```
I was thinking of allowing overrides only for the extensions group and not v1 and that is why I did not put the generic struct, but will do it now.

lavalamp replied:
thanks, that'd be great. I was thinking not so much of the v1 group, but
the new groups we'll make in the future.

On Tue, Oct 13, 2015 at 1:45 PM, Nikhil Jindal <notifications@github.com>
wrote:

> In cmd/kube-apiserver/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/15516#discussion_r41922410>
> :
>
> > @@ -556,7 +565,9 @@ func (s *APIServer) Run(_ []string) error {
> >  		Authorizer:                authorizer,
> >  		AdmissionControl:          admissionController,
> >  		DisableV1:                 disableV1,
> > -		EnableExp:                 enableExp,
> > +		EnableExtensions:          enableExtensions,
>
> Yes I can make this a generic struct:
>
> ResourceOverrides struct {
>   groupVersion string,
>   resources util.Set,
>   isEnableList bool
> }
>
> I was thinking of allowing overrides only for the extensions group and not
> v1 and that is why I did not put the generic struct, but will do it now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15516/files#r41922410>.
>

nikhiljindal replied:
Have added a generic apiGroupVersionOverrides.
But have not added support for enabling/disabling individual v1 resources. I dont think we need it. Can add it later, if required.

erictune commented:
What is the plan for the default value of runtime-config?  Empty string?  Which enables what exactly?

And do we plan to set it to a different value in cluster turnup scripts.

lavalamp commented:
In head, we should use: "v1=true,extensions/v1beta1=true"

In 1.1 branch,
"v1=true,extensions/v1beta1/ingress=true,extensions/v1beta1/horizontalpodautoscaler=true,extensions/v1beta1/job=true"

(Nikhil, does that look right to you?)

Now that I think about it, I think if we set the default in code, we
shouldn't need to do any special config plumbing?

On Tue, Oct 13, 2015 at 2:08 PM, Eric Tune <notifications@github.com> wrote:

> What is the plan for the default value of runtime-config? Empty string?
> Which enables what exactly?
>
> And do we plan to set it to a different value in cluster turnup scripts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15516#issuecomment-147854460>
> .
>

nikhiljindal commented:
Default value is empty string, which enables only the v1 API.
We will set it to ```"runtime-config=v1,extensions/v1beta1,extenstions/v1beta1/jobs=true,extensions/v1beta1/horizontalpodautoscalers=true,extensions/v1beta1/ingress=true"``` in all our turnup scripts.

erictune commented:
@derekwaynecarr trying to start a discussion about what beta groups are on by default in this thread.  Derek just commented [here]( https://github.com/kubernetes/kubernetes/pull/15339#discussion_r41916572) that he prefers beta fetures to be on by default.  I think beta in this context means job, hpa, and ingress.

Propose default runtime-config value is:
 `v1=true,extensions/v1beta1/ingress=true,extensions/v1beta1/job=true,extensions/v1beta1/horizontalPodAutoscaler`.

Does that work?

lavalamp commented:
Why not set the default value in code? That makes me feel better about not
making other people fix their turnup scripts?

On Tue, Oct 13, 2015 at 2:16 PM, Nikhil Jindal <notifications@github.com>
wrote:

> Default value is empty string, which enables only the v1 API.
> We will set it to
> "runtime-config=v1,extensions/v1beta1,extenstions/v1beta1/jobs=true,extensions/v1beta1/horizontalpodautoscalers=true,extensions/v1beta1/ingress=true"
> in all our turnup scripts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15516#issuecomment-147856264>
> .
>

nikhiljindal commented:
Yes @lavalamp what you have LGTM. We will first submit a PR setting the 1.1 values on HEAD, cherrypick into 1.1 and then change the HEAD value.

nikhiljindal commented:
Yes @lavalamp (you keep beating me at posting a reply :))
Agreed that code default is better. Will make that change in this PR.

lavalamp commented:
This plan sounds great, thanks.

On Tue, Oct 13, 2015 at 2:21 PM, Nikhil Jindal <notifications@github.com>
wrote:

> Yes @lavalamp <https://github.com/lavalamp> what you have LGTM. We will
> first submit a PR setting the 1.1 values on HEAD, cherrypick into 1.1 and
> then change the HEAD value.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15516#issuecomment-147857556>
> .
>

nikhiljindal commented:
Pushed a new version with defaults in code and an override structure that is extendable for all api groups rather than extension specific variables.
I think we might need to support "extensions/v1beta1/all=true" for our tests, but havent added it yet.
PTAL

nikhiljindal committed:
Updating swagger spec to include the new extension resources that are enabled by default

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -677,3 +657,60 @@ func (s *APIServer) getRuntimeConfigValue(apiKey string, defaultValue bool) bool
+	// "extensions/v1beta1={true|false} allows users to enable/disable the extensions API.
+	// This takes preference over api/all, if specified.
+	disableExtensions := disableAllAPIs
+	extensionsGroupVersion := "extensions/v1beta1"
```
// TODO: make this a loop over all possible group/versions when there are more of them.

nikhiljindal replied:
Done

lavalamp commented:
> I think we might need to support "extensions/v1beta1/all=true" for our tests, but havent added it yet.

I agree.

lavalamp commented:
This LGTM if tests pass, thanks!

nikhiljindal commented:
more than 95 tests failed on Jenkins with namespace could not be deleted error. Filed https://github.com/kubernetes/kubernetes/issues/15679

Retrying
@k8s-bot ok to test

lavalamp commented:
still LGTM but needs another rebase.

nikhiljindal committed:
Fixing integrations test

nikhiljindal commented:
Rebased and added another commit to fix integration tests.
It is bad that I had to change so many integration tests. They should all be using a single util function to bring up the master. Right now all of them have small differences, which we might be able to get rid of. Will try that in a separate PR.

Hopefully the tests will pass now!

nikhiljindal commented:
Jenkins failed with no image found.
Retrying @k8s-bot ok to test

nikhiljindal commented:
Jenkins failed to fetch git repo.
Retrying @k8s-bot ok to test

nikhiljindal commented:
Deployments e2e is failing because deployments is now turned off by default.
Trying to figure out our salt setup to enable deployments on GCE when ENABLE_DEPLOYMENTS is true.

nikhiljindal committed:
Enabling deployments on GCE when the corresponding env var is set to true

nikhiljindal commented:
Pushed a new commit and the deployment test passed on Jenkins!
@lavalamp PTAL

lavalamp commented:
Can you say why your changes exposed bugs in the integration tests? were we just not running them with experimental stuff on before?

nikhiljindal commented:
Integration tests that were using integration/framework to bring up the master, were testing extensions stuff. But lots of other tests which had their own code to bring up the master were not.

lavalamp commented:
LGTM, thanks!

lavalamp commented:
Please cherrypick this when it merges!

nikhiljindal commented:
Updated the PR description and title to better reflect the current state.

nikhiljindal commented:
@k8s-bot ok to test

nikhiljindal commented:
Self merging for 1.1
